libs/@local/hash-graph-sdk/python/graph_sdk/query.py

```diff
@@ -0,0 +1,352 @@
+from enum import Enum
+from typing import Any, Generic, Never, Protocol, Self, TypeVar, assert_never
+
+from graph_client import QueryToken
+from graph_client.models import (
+    Filter,
+    FilterExpression,
+    FilterExpressionItem,
+    FilterExpressionItem1,
+    FilterItem,
+    FilterItem1,
+    FilterItem2,
+    FilterItem3,
+    FilterItem4,
+    FilterItem5,
+    FilterItem6,
+    FilterItem7,
+)
+from pydantic import BaseModel
+
+
+class FFIConversionProtocol(Protocol):
+    """Protocol for converting to and from FFI objects."""
+
+    def to_ffi(self) -> BaseModel:
+        """Converts the object to a FFI object."""
+        ...
+
+
+class UnaryOperation(Enum):
+    """Unary operations."""
+
+    NOT = "not"
+
+
+class BinaryOperation(Enum):
+    """Binary operations."""
+
+    EQUAL = "equal"
+    NOT_EQUAL = "notEqual"
+    STARTS_WITH = "startsWith"
+    ENDS_WITH = "endsWith"
+    CONTAINS_SEGMENT = "containsSegment"
+
+
+class NaryOperation(Enum):
+    """Nary operations."""
+
+    ALL = "all"
+    ANY = "any"
+
+
+L = TypeVar("L", bound="BaseFilterExpression")
+R = TypeVar("R", bound="BaseFilterExpression")
+F = TypeVar("F", bound="BaseFilter")
+
+
+class BaseFilter(ABC):
+    """Protocol for queries."""
+
+    def all_(
+        self,
+        other: Sequence[F],
+    ) -> "NaryFilter[Self | F]":
+        """Combines two or more queries with an AND."""
+        return NaryFilter(NaryOperation.ALL, [self, *other])
+
+    def __and__(self, other: F) -> "NaryFilter[Self | F]":
+        """Combines two queries with an AND."""
+        if isinstance(other, BaseFilter):
+            return self.all_((other,))
+
+        msg: str  # type: ignore[unreachable]
+        msg = f"unsupported operand type(s) for &: '{type(self)}' and '{type(other)}'"
+        raise TypeError(msg)
+
+    def any_(
+        self,
+        other: Sequence[F],
+    ) -> "NaryFilter[Self | F]":
+        """Combines two or more queries with an OR."""
+        return NaryFilter(NaryOperation.ANY, [self, *other])
+
+    def __or__(self, other: F) -> "NaryFilter[Self | F]":
+        """Combines two queries with an OR."""
+        if isinstance(other, BaseFilter):
+            return self.any_((other,))
+
+        msg: str  # type: ignore[unreachable]
+        msg = f"unsupported operand type(s) for |: '{type(self)}' and '{type(other)}'"
+        raise TypeError(msg)
+
+    def not_(self) -> "UnaryFilter[Self]":
+        """Negates the query."""
+        return UnaryFilter(UnaryOperation.NOT, self)
+
+    def __invert__(self) -> "UnaryFilter[Self]":
+        """Negates the query."""
+        return self.not_()
+
+    @classmethod
+    def always(cls) -> "NaryFilter[Never]":
+        """Returns a query that always returns true."""
+        return NaryFilter(NaryOperation.ALL, [])
+
+    @classmethod
+    def never(cls) -> "NaryFilter[Never]":
+        """Returns a query that always returns false."""
+        return NaryFilter(NaryOperation.ANY, [])
```

Comment:
Why you would ever use this? I'd just delete it.

Reply:
removed in [577146f](https://github.com/hashintel/hash/pull/2868/commits/577146f5ba49ea34e9db3af7ae67a7fc9a47c267)